servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp

```diff
@@ -1473,9 +1476,70 @@ void RenderForwardMobile::_render_shadow_process() {
 void RenderForwardMobile::_render_shadow_end() {
 	RD::get_singleton()->draw_command_begin_label("Shadow Render");
 
-	for (SceneState::ShadowPass &shadow_pass : scene_state.shadow_passes) {
-		RenderListParameters render_list_parameters(render_list[RENDER_LIST_SECONDARY].elements.ptr() + shadow_pass.element_from, render_list[RENDER_LIST_SECONDARY].element_info.ptr() + shadow_pass.element_from, shadow_pass.element_count, shadow_pass.flip_cull, shadow_pass.pass_mode, shadow_pass.rp_uniform_set, scene_shader.default_specialization, false, Vector2(), shadow_pass.lod_distance_multiplier, shadow_pass.screen_mesh_lod_threshold, 1, shadow_pass.element_from);
-		_render_list_with_draw_list(&render_list_parameters, shadow_pass.framebuffer, shadow_pass.clear_depth ? RD::DRAW_CLEAR_DEPTH : RD::DRAW_DEFAULT_ALL, Vector<Color>(), 0.0f, 0, shadow_pass.rect);
+	if (RenderingDeviceCommons::render_pass_opts_enabled) {
+		// Render all cascades in the same pass since they render on different regions
+		uint32_t curr_shadow_index = 0;
+
+		while (curr_shadow_index < scene_state.shadow_passes.size()) {
+			uint32_t pass_amount = 0;
+			SceneState::ShadowPass start_shadow_pass = scene_state.shadow_passes[curr_shadow_index];
+			switch (start_shadow_pass.light_type) {
+				case RS::LIGHT_DIRECTIONAL: {
+					uint32_t i = 0;
+					while (i < scene_state.shadow_passes.size() && scene_state.shadow_passes[i].light_type == RS::LIGHT_DIRECTIONAL) {
+						i++;
+						pass_amount++;
+					}
+					break;
+				}
+				// A cubemap is made of 6 faces
+				case RS::LIGHT_OMNI:
+					pass_amount = 6;
```

Comment:
Basically, Spotlights + Dual Parabaloid omni lights need the same treatment as DirectionalLights.

The overall flow of shadow rendering is like this
```
- For each face of each omnilight
    - `_render_shadow_append()`
    - `if face % 6 ==5`
         - `_render_shadow_end()`
         - Copy to positional shadow atlas
- Clear the Directional Shadow map
- For each directional light
    - for each split
        - `_render_shadow_append()`
- For each spotlight and each DP omnilight face
    - `_render_shadow_append()`
- `_render_shadow_end()` // Renders all Directional Lights directly to directional atlas, renders all spotlights and DPs to positional atlas
```
So Cube Omnilights will need to have one draw list for each face, but spotlights and DPs should be able to all share a draw list after Directional Lights have finished. 

Currently, it works this awkward way because each light is responsible for clearing its own section. So maybe we should change the order and do Spotlights and DPs first and cubes after. Then we can open a draw list with a full clear, render all spots and DPs. Close the draw list. Then render all cube omnilights and copy them over. What do you think?

Reply:
OK, I've reviewed these remarks.

Well spotted! It was "working fine" because the loop later does:

```
for (uint32_t j = 0; j < pass_amount && curr_shadow_index + j < scene_state.shadow_passes.size(); j++) {
}
```

And the `curr_shadow_index + j < scene_state.shadow_passes.size()` part was hiding the bug from me during execution, as it was early-outing the loop before pass_amount >= 2. But this would've certainly break on other projects.

I just fixed this bug.

> Currently, it works this awkward way because each light is responsible for clearing its own section. So maybe we should change the order and do Spotlights and DPs first and cubes after.

This is problematic because quadrants may not be updated if nothing changed. For what you want to do to work, we would need to do the following:

Let's say we have 4 spotlights, 512x512 each:

1. Clear 1024x1024 region
2. SetViewport to the first 512x512 block
3. Render
4. Go to step for the next 512x512 until all 4 blocks are rendered
5. Close draw list

The problem with that, is that if a single spotlight in those 4 blocks doesn't want to be updated, we must still update it or else the 1024x1024 clear will erase the 512x512 region.

So either we're extremely smart in how we group lights (e.g. clear & render to a 1024x512 region, then clear & render to a 512x512 region), or we force lights to update even if they don't have to if they're caught in the middle (or copy those lights to a temporary location so we can then restore their contents).

It would complicate the logic behind it, a lot.

Reply:
Oooooh, right. I forgot how aggressively we cache shadows. Indeed, ideally we only update a small subset of lights per frame, so we shouldn't clear the entire atlas.